File: compress.py
import os
from PIL import Image

# Directory containing the images to compress
assets_dir = os.path.join(os.path.dirname(__file__), "src", "assets")

def compress_image(file_path):
    old_size = os.path.getsize(file_path)
    filename = os.path.basename(file_path)
    
    # We only want to compress files larger than 150KB
    if old_size < 150 * 1024:
        print(f"Skipping {filename} ({old_size / 1024:.1f} KB) - already small")
        return
        
    try:
        with Image.open(file_path) as img:
            width, height = img.size
            max_dim = 1200
            
            # Resize if the image is very large (e.g., camera/generated outputs)
            if width > max_dim or height > max_dim:
                if width > height:
                    new_width = max_dim
                    new_height = int(height * (max_dim / width))
                else:
                    new_height = max_dim
                    new_width = int(width * (max_dim / height))
                
                # Resize with high quality Lanczos filter
                img = img.resize((new_width, new_height), Image.Resampling.LANCZOS)
                print(f"Resized {filename} from {width}x{height} to {new_width}x{new_height}")
            
            ext = os.path.splitext(file_path)[1].lower()
            
            if ext == '.png':
                # Convert to palette mode (indexed color) for massive PNG size reduction
                # This keeps the file as a valid PNG and preserves transparency (alpha)
                if img.mode in ('RGBA', 'LA') or (img.mode == 'P' and 'transparency' in img.info):
                    rgba = img.convert('RGBA')
                    quantized = rgba.quantize(colors=256, method=Image.Quantize.MAXCOVERAGE)
                    quantized.save(file_path, 'PNG', optimize=True)
                else:
                    rgb = img.convert('RGB')
                    quantized = rgb.quantize(colors=256, method=Image.Quantize.MAXCOVERAGE)
                    quantized.save(file_path, 'PNG', optimize=True)
            elif ext in ('.jpg', '.jpeg'):
                # Save JPEG with 80% quality and progressive/optimized rendering
                rgb = img.convert('RGB')
                rgb.save(file_path, 'JPEG', quality=80, optimize=True)
            else:
                print(f"Unsupported extension for {filename}: {ext}")
                return
                
        new_size = os.path.getsize(file_path)
        reduction = (old_size - new_size) / old_size * 100
        print(f"Compressed {filename}: {old_size / 1024 / 1024:.2f} MB -> {new_size / 1024:.1f} KB ({reduction:.1f}% reduction)")
    except Exception as e:
        print(f"Error compressing {filename}: {str(e)}")

def main():
    print(f"Scanning directory: {assets_dir}")
    if not os.path.exists(assets_dir):
        print("Error: src/assets directory not found.")
        return
        
    for file in os.listdir(assets_dir):
        file_path = os.path.join(assets_dir, file)
        if os.path.isfile(file_path):
            ext = os.path.splitext(file)[1].lower()
            if ext in ('.png', '.jpg', '.jpeg'):
                compress_image(file_path)
    print("All operations complete.")

if __name__ == "__main__":
    main()

File: compress.js
import fs from 'fs';
import path from 'path';
import { fileURLToPath } from 'url';
import sharp from 'sharp';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

const assetsDir = path.join(__dirname, 'src', 'assets');

async function compressImage(filePath) {
  const oldSize = fs.statSync(filePath).size;
  const filename = path.basename(filePath);

  // Skip files already under 150KB
  if (oldSize < 150 * 1024) {
    console.log(`Skipping ${filename} (${(oldSize / 1024).toFixed(1)} KB) - already small`);
    return;
  }

  try {
    const ext = path.extname(filePath).toLowerCase();
    const tempPath = filePath + '.tmp';

    let pipeline = sharp(filePath);
    const metadata = await pipeline.metadata();

    const maxDim = 1200;
    let width = metadata.width;
    let height = metadata.height;

    // Resize if too large
    if (width > maxDim || height > maxDim) {
      if (width > height) {
        height = Math.round(height * (maxDim / width));
        width = maxDim;
      } else {
        width = Math.round(width * (maxDim / height));
        height = maxDim;
      }
      pipeline = pipeline.resize(width, height, { fit: 'inside' });
      console.log(`Resizing ${filename} to ${width}x${height}`);
    }

    // Apply format-specific compression
    if (ext === '.png') {
      // Use palette-based compression (indexed color) for massive PNG size reduction
      pipeline = pipeline.png({ palette: true, quality: 80, effort: 6 });
    } else if (ext === '.jpg' || ext === '.jpeg') {
      pipeline = pipeline.jpeg({ quality: 80, progressive: true });
    } else {
      console.log(`Unsupported extension for ${filename}: ${ext}`);
      return;
    }

    await pipeline.toFile(tempPath);

    // Replace original file with the compressed one
    fs.renameSync(tempPath, filePath);

    const newSize = fs.statSync(filePath).size;
    const reduction = ((oldSize - newSize) / oldSize) * 100;
    console.log(`Compressed ${filename}: ${(oldSize / 1024 / 1024).toFixed(2)} MB -> ${(newSize / 1024).toFixed(1)} KB (${reduction.toFixed(1)}% reduction)`);
  } catch (err) {
    console.error(`Error compressing ${filename}:`, err.message);
  }
}

async function main() {
  console.log(`Scanning directory: ${assetsDir}`);
  if (!fs.existsSync(assetsDir)) {
    console.error('Error: src/assets directory not found.');
    return;
  }

  const files = fs.readdirSync(assetsDir);
  for (const file of files) {
    const filePath = path.join(assetsDir, file);
    if (fs.statSync(filePath).isFile()) {
      const ext = path.extname(file).toLowerCase();
      if (['.png', '.jpg', '.jpeg'].includes(ext)) {
        await compressImage(filePath);
      }
    }
  }
  console.log('All operations complete.');
}

main();
